### PR TITLE
Add a `SocketPath` type for `linux` systems

### LINKED ISSUE #10376: Pydantic FilePath does not accept sockets
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

I want to pass a file path to an existing unix domain socket to a pydantic class and have it validate that the path exists.
from linux's perspective, its the same as any other file.
either it exists or it doesn't.

Pydantic seems to draw the distinction, actually looking at the type of the file object, and sees that its a socket and not a concrete file.

Strangely, Pydantic accepts symlinks to real files even though symlinks themselves are not real files either :stuck_out_tongue: 
```
>       Reproduction(host="/tmp/tmp_socket")  # explodes
E       pydantic_core._pydantic_core.ValidationError: 1 validation error for Reproduction
E       host
E         Path does not point to a file [type=path_not_file, input_value='/tmp/tmp_socket', input_type=str]
```

### Example Code

```Python
def test_pydantic_socket_validation():
    from pathlib import Path
    from pydantic import BaseModel, FilePath
    import socket


    class Reproduction(BaseModel):
        host: FilePath

    # clean up after ourselves
    socket_path = Path("/tmp")/"tmp_socket"
    socket_path.unlink(missing_ok=True)
    # create a socket
    sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    sock.bind("/tmp/tmp_socket")

    # create the real files for comparison
    real_file = Path("/tmp") / "foo.txt"
    real_file.unlink(missing_ok=True)  # clean up after ourselves
    real_file.touch()
    link_to_real_file = Path("/tmp") / "foo.lnk.txt"
    link_to_real_file.unlink(missing_ok=True)  # clean up after ourselves
    link_to_real_file.symlink_to(real_file)
    # assert real path to real file works
    Reproduction(host="/tmp/foo.txt")
    # assert symlinked path to real file works
    Reproduction(host="/tmp/foo.lnk.txt")
    # assert real path to socket works
    Reproduction(host="/tmp/tmp_socket")  # explodes
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.9.1
        pydantic-core version: 2.23.3
          pydantic-core build: profile=release pgo=false
                 install path: /PATH_REDACTED/.venv/lib/python3.12/site-packages/pydantic
               python version: 3.12.5 (main, Aug 14 2024, 05:08:31) [Clang 18.1.8 ]
                     platform: Linux-6.5.0-44-generic-x86_64-with-glibc2.35
             related packages: typing_extensions-4.12.2
                       commit: unknown
```

### PR DESCRIPTION
**Pull request by theunkn0wn1**
## Change Summary

The existing FilePath type does not accept sockets, as they fail the `Path.is_file(self)` check.
As such, I added a new `SocketPath` type and corresponding validator using `Path.is_socket(self)`.

## Related issue number
fix #10376

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @sydney-runkle

### CODE REVIEW

**Comment by theunkn0wn1** [comment]
Test will need to be skipped on windows; as it doesn't have sockets to begin with.
Not sure why the tests are failing on mac.

File: pydantic/types.py [hunk 1]
```diff
@@ -1238,7 +1239,7 @@ class Model(BaseModel):
 
 @_dataclasses.dataclass
 class PathType:
-    path_type: Literal['file', 'dir', 'new']
+    path_type: Literal['file', 'dir', 'new', 'socket']
```
Comment (Viicos) [inline pydantic/types.py:1242]:
```suggestion
    if sys.platform != "win32":
        path_type: Literal['file', 'dir', 'new', 'socket']
    else:
        path_type: Literal['file', 'dir', 'new']
```

**Review by Viicos — COMMENTED** [review]
I'm wondering if we should do something like:

Edit: Seems like on Windows, `is_socket` returns `False`, so maybe it's safe to not add these kind of conditions.

**Comment by sydney-runkle** [comment]
Looking good though, happy to support this once we get tests passing, etc :).

**Comment by theunkn0wn1** [comment]
ok for the issue with the Mac tests - pytest seems to be configured wrong, the temporary paths its generating with its fixture exceed the maximum length of a socket name on that platform.
Only way i can see to fix that would be to change the pytest config to use a more sensible base directory.
the trouble? i don't have a mac to test with and don't know what a sensible base path for that platform is.
Can I ask for some pointers here?

File: pydantic/types.py (hunk 1 above)
Comment (theunkn0wn1) [reply pydantic/types.py:1242]: i am suprised that conditional attributes on dataclasses are a thing

Comment (Viicos) [reply pydantic/types.py:1242]: But anyway, turns out you can also have sockets on Windows, so let's not add any conditions.

**Comment by Viicos** [comment]
> ok for the issue with the Mac tests - pytest seems to be configured wrong, the temporary paths its generating with its fixture exceed the maximum length of a socket name on that platform.
> Only way i can see to fix that would be to change the pytest config to use a more sensible base directory.
> the trouble? i don't have a mac to test with and don't know what a sensible base path for that platform is.
> Can I ask for some pointers here?

I'm ok only running the test on Linux. The goal isn't to test that creating sockets works, but to ensure that `validate_socket` works as expected when encountering a socket path.

File: pydantic/types.py (hunk 1 above)
Comment (theunkn0wn1) [reply pydantic/types.py:1242]: Interesting, ok. Closing these conversations then.

**Comment by theunkn0wn1** [comment]
Added condition to skip new test if not linux.
squashed commits to one.
rebased against tip of upstream main. :)

**Comment by theunkn0wn1** [comment]
please review :)

**Review by sydney-runkle — APPROVED** [review]
Great, thanks!

**Comment by theunkn0wn1** [comment]
@sydney-runkle when can i expect this change to be released? :innocent:

**Comment by sydney-runkle** [comment]
@theunkn0wn1, with our 2.10 release in late October! Feel free to install from main in the meantime :)

**Pull request merged**